jerry-core/jerry-snapshot.c

```diff
@@ -630,3 +632,409 @@ jerry_exec_snapshot (const void *snapshot_p, /**< snapshot */
+    const uint8_t *str_end_p = str_buffer_p + str_buffer_size;
+
+    result = true;
+
+    while (str_start_p < str_end_p)
+    {
+      if (!lit_char_is_identifier_part (str_start_p))
+      {
+        result = false;
+        break;
+      }
+      lit_utf8_incr (&str_start_p);
+    }
+  }
+  else
+  {
+    result = false;
+  }
+
+  ECMA_FINALIZE_UTF8_STRING (str_buffer_p, str_buffer_size);
+
+  return result;
+} /* ecma_string_is_valid_identifier */
+
+#endif /* JERRY_ENABLE_SNAPSHOT_SAVE */
+
+/**
+ * Copy certain string literals into the given buffer in a specified format,
+ * which are valid identifiers and none of them are magic string.
+ *
+ * @return size of the literal-list in bytes, at most equal to the buffer size,
+ *          if the source parsed successfully and the list of the literals isn't empty,
+ *         0 - otherwise.
+ */
+size_t
+jerry_parse_and_save_literals (const jerry_char_t *source_p, /**< script source */
+                               size_t source_size, /**< script source size */
+                               bool is_strict, /**< strict mode */
+                               uint8_t *buffer_p, /**< buffer to save literals to */
+                               size_t buffer_size, /**< the buffer's size */
+                               bool is_c_format) /**< format-flag */
+{
+#ifdef JERRY_ENABLE_SNAPSHOT_SAVE
+  ecma_value_t parse_status;
+  ecma_compiled_code_t *bytecode_data_p;
+  parse_status = parser_parse_script (source_p,
+                                      source_size,
+                                      is_strict,
+                                      &bytecode_data_p);
+
+  if (ECMA_IS_VALUE_ERROR (parse_status))
+  {
+    ecma_free_value (parse_status);
+    return 0;
+  }
+
+  ecma_free_value (parse_status);
+  ecma_bytecode_deref (bytecode_data_p);
+
+  ecma_lit_storage_item_t *string_list_p = JERRY_CONTEXT (string_list_first_p);
+  lit_utf8_size_t literal_count = 0;
+  size_t total_bytes = 0;
+
+  /* Count the valid and non-magic identifiers in the list */
+  while (string_list_p != NULL)
+  {
+    for (int i = 0; i < ECMA_LIT_STORAGE_VALUE_COUNT; i++)
+    {
+      if (string_list_p->values[i] != JMEM_CP_NULL)
+      {
+        ecma_string_t *literal_p = JMEM_CP_GET_NON_NULL_POINTER (ecma_string_t,
+                                                                 string_list_p->values[i]);
+        /* We don't save a literal which isn't a valid identifier
+           or it's a magic string */
+        if (ecma_string_is_valid_identifier (literal_p)
+            && ecma_get_string_magic (literal_p) == LIT_MAGIC_STRING__COUNT)
+        {
+          literal_count++;
+        }
+      }
+    }
+
+    string_list_p = JMEM_CP_GET_POINTER (ecma_lit_storage_item_t, string_list_p->next_cp);
+  }
+
+  if (literal_count == 0)
+  {
+    return 0;
+  }
+
+  ecma_string_t *literal_array[literal_count];
+  ecma_string_t *literal_array_sorted[literal_count];
```

Comment:
I would change these to real allocation, since stack consumption might be too big.